docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,833 @@
+purpose is to determine where to read and create a File System reader that can only read that
+section of the file.
+
+At the domain level, reading goes through three phases:
+1. Using commands such as `Seek`, `Next`, or `Prev`, find the position of the domain of
+interest that contains the time range we wish to read.
+2. Load the pointer at the found position into the iterator as its `value` field.
+3. Create a section reader based on information in the stored pointer: the offset for
+this reader is set to 0 where the domain starts in the file system, and EOF where the
+length of the domain is reached.
+4. Use the section reader as an `ReaderAtCloser` to read binary data.
+
+Note that there are various race conditions here! These race conditions are a
+combination of three categories: change to the pointer itself, change to the domain index, and
+change to the underlying data.
+
+At step 2, changes made to the domain index prevent `i.position` from pointing to the right
+pointer; at step 3, changes made to the pointer prevent creating the right reader; at
+step 4, changes made to file prevent reading the correct data. Here is a diagram that
+summarizes these races:
+
+<div align="center">
+<img width="90%" src="img/0019-240529-cesium-race-conditions/Read-Races.png">
+    <h6>Life cycle of a domain iterator</h6>
+</div>
+
+#### 4.2.2.1 Race conditions involving the domain index
+If the domain index's content was changed between finding the position of the pointer of
+interest and reading it, the iterator will not read the correct pointer. For example, assume the
+seeked position of interest is 3. These changes may lead to reading the incorrect pointer:
+- `Write` inserting a pointer before the position of interest: the pointer of interest
+is found at position 6 instead of 5.
+- `Delete` deleting pointer(s) before the position of interest: the pointer of interest
+is found at position 4 instead of 5.
+- `Delete` deletes the pointer at the position of interest: the pointer of interest is
+no longer in the domain index, instead we would be reading what was at position 6.
+
+Notice that changes to the domain index after the position of interest does not invalidate the
+position.
+
+Also note that not only do these changes invalidate loading the pointer into the iterator,
+they also invalidate future iterator movements – the position of the iterator is no
+longer consistent with the value.
+
+#### 4.2.2.2 Race conditions involving the pointer
+Note that for creation of the reader, the offset and length information are not based on
+the pointer stored in memory, but the pointer loaded into the iterator. If the
+pointer of interest was modified, the created read does not read the correct chunk
+of data.
+
+This is not necessarily a problem – as documented, the iterator does not iterate over
+pointers in a snapshot of the DB, but rather stores a pointer that was correct
```

Comment:
Write "database" instead of "DB". Also, be clear what database you are reffering to.